### PR TITLE
Reland (2): "Fix how Gradle resolves Android plugin"

### LINKED ISSUE #141940: `error: package io.flutter.plugins.* does not exist` when running app on Flutter beta/master channel
### Steps to reproduce

add shared_preferences using flutter pub add or directly using pubspec.yaml (tried with 2.2.2 and 2.2.1)

### Expected results

App launches

### Actual results

Build fails with exit code 1
                                                                             ^

### Code sample

<details open><summary>Code sample</summary>

N/A
</details>

### Screenshots or Video

<details open>
<summary>Screenshots / Video demonstration</summary>

[Upload media here]

</details>

### Logs

<details open><summary>Logs</summary>

```console
Launching lib/main.dart on SM G975W in debug mode...
/Users/dany/Dev/flutter-projects/mobile-publisher/android/app/src/main/java/io/flutter/plugins/GeneratedPluginRegistrant.java:84: error: package io.flutter.plugins.sharedpreferences does not exist
      flutterEngine.getPlugins().add(new io.flutter.plugins.sharedpreferences.SharedPreferencesPlugin());
                                                                             ^
1 error

FAILURE: Build failed with an exception.

* What went wrong:
Execution failed for task ':app:compileDebugJavaWithJavac'.
> Compilation failed; see the compiler error output for details.

* Try:
> Run with --stacktrace option to get the stack trace.
> Run with --info or --debug option to get more log output.
> Run with --scan to get full insights.

* Get more help at https://help.gradle.org

BUILD FAILED in 6s
Error: Gradle task assembleDebug failed with exit code 1


Exited (1).
```

</details>

### Flutter Doctor output

<details open><summary>Doctor output</summary>

```console
[✓] Flutter (Channel beta, 3.19.0-0.2.pre, on macOS 14.2.1 23C71 darwin-arm64, locale en-CA)
    • Flutter version 3.19.0-0.2.pre on channel beta at /Users/dany/Dev/flutter
    • Upstream repository https://github.com/flutter/flutter.git
    • Framework revision c1df7d07ac (3 days ago), 2024-01-17 15:42:07 -0800
    • Engine revision 4f18bb4dcb
    • Dart version 3.3.0 (build 3.3.0-279.0.dev)
    • DevTools version 2.31.0

[✓] Android toolchain - develop for Android devices (Android SDK version 32.0.0)
    • Android SDK at /Users/dany/Library/Android/sdk
    • Platform android-33, build-tools 32.0.0
    • Java binary at: /Applications/Android Studio 2.app/Contents/jbr/Contents/Home/bin/java
    • Java version OpenJDK Runtime Environment (build 17.0.7+0-17.0.7b1000.6-10550314)
    • All Android licenses accepted.

[✓] Xcode - develop for iOS and macOS (Xcode 15.2)
    • Xcode at /Applications/Xcode.app/Contents/Developer
    • Build 15C500b
    • CocoaPods version 1.14.3

[✓] Chrome - develop for the web
    • Chrome at /Applications/Google Chrome.app/Contents/MacOS/Google Chrome

[✓] Android Studio (version 2021.3)
    • Android Studio at /Applications/Android Studio.app/Contents
    • Flutter plugin can be installed from:
      🔨 https://plugins.jetbrains.com/plugin/9212-flutter
    • Dart plugin can be installed from:
      🔨 https://plugins.jetbrains.com/plugin/6351-dart
    • Java version OpenJDK Runtime Environment (build 11.0.13+0-b1751.21-8125866)

[✓] Android Studio (version 2023.1)
    • Android Studio at /Applications/Android Studio 2.app/Contents
    • Flutter plugin can be installed from:
      🔨 https://plugins.jetbrains.com/plugin/9212-flutter
    • Dart plugin can be installed from:
      🔨 https://plugins.jetbrains.com/plugin/6351-dart
    • Java version OpenJDK Runtime Environment (build 17.0.7+0-17.0.7b1000.6-10550314)

[✓] VS Code (version 1.85.1)
    • VS Code at /Applications/Visual Studio Code.app/Contents
    • Flutter extension version 3.80.0

[✓] VS Code (version 1.82.2)
    • VS Code at /Applications/Visual Studio Code copy.app/Contents
    • Flutter extension version 3.80.0

[✓] Connected device (3 available)
    • SM G975W (mobile) • R38M3066HET • android-arm64  • Android 12 (API 31)
    • macOS (desktop)   • macos       • darwin-arm64   • macOS 14.2.1 23C71 darwin-arm64
    • Chrome (web)      • chrome      • web-javascript • Google Chrome 120.0.6099.234

[✓] Network resources
    • All expected network resources are available.

• No issues found![Paste your output here]
```

</details>

### LINKED ISSUE #142487: [Android] Re-land `flutter.groovy` logic needed to support Gradle Kotlin DSL
https://github.com/flutter/flutter/pull/142464 reverts logic in `flutter.groovy` in order to fix https://github.com/flutter/flutter/issues/141940, but in the process, has removed some logic required to add support for the Gradle Kotlin DSL that was added in https://github.com/flutter/flutter/pull/140744. This issue serves as a reminder to re-land this logic to maintain that support.

This involves two steps:

- [ ] Replace the [old implementation](https://github.com/flutter/flutter/pull/142464/files#diff-f0cdaf8963c97a175f64fe0fad130260d0614dd6cecd5cbe9580628bd51c5a4cR767-R770) of `doesSupportAndroidPlatform` with the [new implementation](https://github.com/flutter/flutter/pull/140744/files#diff-f0cdaf8963c97a175f64fe0fad130260d0614dd6cecd5cbe9580628bd51c5a4cR458-R469) created in https://github.com/flutter/flutter/pull/140744.
- [ ]  When https://github.com/flutter/flutter/pull/137115 is re-landed, add back logic that [checks for Gradle settings configuration](https://github.com/flutter/flutter/pull/140744/files#diff-f0cdaf8963c97a175f64fe0fad130260d0614dd6cecd5cbe9580628bd51c5a4cR476-R486) for both Groovy or Kotlin

### PR DESCRIPTION
**Pull request by Gustl22**
Previous PR: #137115, 
Revert: #142464
Fixes #141940
Closes #142487

## Pre-launch Checklist

- [x] I read the [Contributor Guide] and followed the process outlined there for submitting PRs.
- [x] I read the [Tree Hygiene] wiki page, which explains my responsibilities.
- [x] I read and followed the [Flutter Style Guide], including [Features we expect every widget to implement].
- [x] I signed the [CLA].
- [x] I listed at least one issue that this PR fixes in the description above.
- [x] I updated/added relevant documentation (doc comments with `///`).
- [x] I added new tests to check the change I am making, or this PR is [test-exempt].
- [x] All existing and new tests are passing.

If you need help, consider asking for advice on the #hackers-new channel on [Discord].

[Contributor Guide]: https://github.com/flutter/flutter/wiki/Tree-hygiene#overview
[Tree Hygiene]: https://github.com/flutter/flutter/wiki/Tree-hygiene
[test-exempt]: https://github.com/flutter/flutter/wiki/Tree-hygiene#tests
[Flutter Style Guide]: https://github.com/flutter/flutter/wiki/Style-guide-for-Flutter-repo
[Features we expect every widget to implement]: https://github.com/flutter/flutter/wiki/Style-guide-for-Flutter-repo#features-we-expect-every-widget-to-implement
[CLA]: https://cla.developers.google.com/
[flutter/tests]: https://github.com/flutter/tests
[breaking change policy]: https://github.com/flutter/flutter/wiki/Tree-hygiene#handling-breaking-changes
[Discord]: https://github.com/flutter/flutter/wiki/Chat

### CODE REVIEW

**Comment by Gustl22** [comment]
Tests work as expected for enabled gradle daemon (before the fix): https://github.com/flutter/flutter/pull/142498/checks?check_run_id=21166626702

**Comment by Gustl22** [comment]
CC @bartekpacia @camsim99 @reidbaker

File: packages/flutter_tools/gradle/src/main/groovy/flutter.groovy [hunk 1]
```diff
@@ -67,7 +66,7 @@ class FlutterExtension {
      * Specifies the relative directory to the Flutter project directory.
      * In an app project, this is ../.. since the app's Gradle build file is under android/app.
      */
-    String source
+    String source = "../.."
```
Comment (reidbaker) [inline packages/flutter_tools/gradle/src/main/groovy/flutter.groovy:69]: Is there a test that covers when source is not set that we can still build?

File: packages/flutter_tools/gradle/src/main/groovy/flutter.groovy [hunk 2]
```diff
@@ -620,19 +624,104 @@ class FlutterPlugin implements Plugin<Project> {
...
+        }
+        List<Map<String, Object>> deps = getPluginDependencies(project)
+        List<String> plugins = getPluginList(project).collect { it.name as String }
+        deps.removeIf { plugins.contains(it.name) }
+        deps.each {
+            Project pluginProject = project.rootProject.findProject(":${it.name}")
+            if (pluginProject == null) {
+                // Plugin was not included in `settings.gradle`, but is listed in `.flutter-plugins`.
+                project.logger.error("Plugin project :${it.name} listed, but not found. Please fix your settings.gradle/settings.gradle.kts.")
+            } else if (doesSupportAndroidPlatform(pluginProject.projectDir.parentFile.path as String)) {
```
Comment (reidbaker) [inline packages/flutter_tools/gradle/src/main/groovy/flutter.groovy:673]: Should this line warn the caller that their project is misconfigured?

**Review by reidbaker — COMMENTED** [review]
This looks good and I appreciate the effort you took to write a test that overrides the daemon options to catch a regression. 

I am going to give this another pass after I have had some time to check it out and build locally but it looks good.

File: packages/flutter_tools/gradle/src/main/groovy/flutter.groovy (hunk 1 above)
Comment (Gustl22) [reply packages/flutter_tools/gradle/src/main/groovy/flutter.groovy:69]: I'm was not sure if the test should be part of this change, as I don't propose to remove setting the source in `app/build.gradle`.

There are also two cases:
Having unset the source in `app/build.gradle` 

```
flutter {
    source ''
}
```

or having non source set at all in `app/build.gradle`

```
flutter {
}
```

https://github.com/flutter/flutter/blob/cbc18b2c51ca27b158976ff6cf9a429ff5bffa3f/packages/flutter_tools/templates/app_shared/android-java.tmpl/app/build.gradle.tmpl#L56

One can argue, that the build works without setting the source with this change and we should ensure it's working in the future, if that's what we actually want. I'll add a test and then can see if this fits the requirements.

File: packages/flutter_tools/gradle/src/main/groovy/flutter.groovy (hunk 2 above)
Comment (Gustl22) [reply packages/flutter_tools/gradle/src/main/groovy/flutter.groovy:673]: Indeed, an example is tested  [here](https://github.com/flutter/flutter/pull/142498/files#diff-2a75005f6853f1c4d2c511ea6ed3b51ddeac0753a7353c4a675bd4ba5f4cb104R154)

File: packages/flutter_tools/gradle/src/main/groovy/flutter.groovy (hunk 1 above)
Comment (bartekpacia) [reply packages/flutter_tools/gradle/src/main/groovy/flutter.groovy:69]: I think this change belongs in a separate PR

Comment (Gustl22) [reply packages/flutter_tools/gradle/src/main/groovy/flutter.groovy:69]: I added a test file: android_gradle_source_path_test

Comment (Gustl22) [reply packages/flutter_tools/gradle/src/main/groovy/flutter.groovy:69]: > I think this change belongs in a separate PR

Oh ok, my page didn't update to your respone. So I can add another PR just to change the default...

Comment (Gustl22) [reply packages/flutter_tools/gradle/src/main/groovy/flutter.groovy]: Moved changes to #142934

**Comment by Gustl22** [comment]
@reidbaker small ping if there's time to finish the review :)

**Comment by reidbaker** [comment]
I added a teammate who can give this a faster review. Sorry for the delay.

**Comment by camsim99** [comment]
> Tests fail as expected for enabled gradle daemon (before the fix): https://github.com/flutter/flutter/pull/142498/checks?check_run_id=21166626702
> 
> I reverted the revert with [309767c](https://github.com/flutter/flutter/commit/309767ce37cfc966bd153a2e4b9720795057d599) and added the tests with [851fe0d](https://github.com/flutter/flutter/commit/851fe0de9bf3e6c0f70c859ff6fff7006ceca34c) and added the fix with [cddc017](https://github.com/flutter/flutter/commit/cddc01714d0a1d6b25788a0347ab2e2b603480d7)

The test added LGTM but can you add some context as to why [cddc017](https://github.com/flutter/flutter/commit/cddc01714d0a1d6b25788a0347ab2e2b603480d7)
 fixes https://github.com/flutter/flutter/issues/141940?

Also pinged @gmackall for a review if you can since you are more of a Gradle expert!

**Comment by Gustl22** [comment]
Thank you :)

> The test added LGTM but can you give some context as to why [cddc017](https://github.com/flutter/flutter/commit/cddc01714d0a1d6b25788a0347ab2e2b603480d7)
fixes https://github.com/flutter/flutter/issues/141940?

As stuartmorgan described in https://github.com/flutter/flutter/issues/141940#issuecomment-1912645081 a Singleton persists across builds in gradle daemon. So the Singleton keyword is removed in cddc01714d0a1d6b25788a0347ab2e2b603480d7. To not read the same file twice during the same build, some local variables were introduced (`pluginList` and `pluginDependencies`) in flutter.groovy to ensure consistency and push performance.

**Review by camsim99 — APPROVED** [review]
The fix for the cause of the revert LGTM! @bartekpacia or @gmackall for second review?

**Comment by gmackall** [comment]
I would defer to someone who reviewed the original PR (I think the delta here between original and reland is small) but can take a look if no one else is able to. It'll take me a bit though, as I'll have to review (original+reland delta)

**Review by reidbaker — APPROVED** [review]
Thank you for your effort to reland this pr.

**Comment by Gustl22** [comment]
Thank you for reviewing!

**Pull request merged**